internal/pkg/agent/install/uninstall.go

```diff
@@ -235,9 +235,15 @@ func applyDynamics(ctx context.Context, log *logger.Logger, cfg *config.Config)
 		varsArray := make([]*transpiler.Vars, 0)
 		var wg sync.WaitGroup
 		wg.Add(1)
+
+		ctx, cancel := context.WithCancel(ctx)
+
+		// The composable system will continuously run, we are only interested in the first run on of the
+		// renderer to collect the variables we should stop the execution.
 		varsCallback := func(vv []*transpiler.Vars) {
 			varsArray = vv
 			wg.Done()
+			cancel()
```

Comment:
imho this changes the behavior and introduces potential race that when the callback can be called before the context cancellation is detected in the caller.

and the ```cancel``` is not called here if the callback is not called for some reason, might be flagged by linter. the pattern is usually to guarantee the ```defer cancel()```

think using ```sync.Once``` might safer fix if you want to guarantee that you only need to capture the first run.

maybe for example:
https://go.dev/play/p/-KpLGWS03yX


what do you think about something like this?
```
		var (
			wg sync.WaitGroup
			once sync.Once
		)
		wg.Add(1)
		ctx, cancel := context.WithCancel(ctx)
		defer cancel()
		varsCallback := func(vv []*transpiler.Vars) {
			once.Do(func() {
				varsArray = vv
				wg.Done()
				cancel()
			})
		}

		ctrl, err := composable.New(log, cfg)
		if err != nil {
			return nil, err
		}
		_ = ctrl.Run(ctx, varsCallback)
		wg.Wait()
```

Reply:
another possible problem here is if the callback is never called due to any error in Run (the error is not checked now), then it will wait on WaitGroup at the end forever

Reply:
I like this approach, it's less prone to problems

Reply:
I initially though about going with `Once` before doing the `cancel` thing, I thought it was adding more locking complexity over the current scenario, since we only want to opt out on the first render were we would assume we have all the appropriate information collected. But I do think a mix of cancelling the ctx and waiting on the `Done()` would be better than mixing cancel with the current Waitgroup, I will make the change.